src/platforms/wayland/displayclient.cpp

```diff
@@ -0,0 +1,856 @@
+    auto output = static_cast<Output*>(data);
+
+    auto& dcout = output->dcout;
+    dcout.top_left = {x, y};
+    dcout.physical_size_mm = {physical_width, physical_height};
+
+    switch (subpixel)
+    {
+    case WL_OUTPUT_SUBPIXEL_UNKNOWN:
+        dcout.subpixel_arrangement = MirSubpixelArrangement::mir_subpixel_arrangement_unknown;
+        break;
+
+    case WL_OUTPUT_SUBPIXEL_NONE:
+        dcout.subpixel_arrangement = MirSubpixelArrangement::mir_subpixel_arrangement_none;
+        break;
+
+    case WL_OUTPUT_SUBPIXEL_HORIZONTAL_RGB:
+        dcout.subpixel_arrangement = MirSubpixelArrangement::mir_subpixel_arrangement_horizontal_rgb;
+        break;
+
+    case WL_OUTPUT_SUBPIXEL_HORIZONTAL_BGR:
+        dcout.subpixel_arrangement = MirSubpixelArrangement::mir_subpixel_arrangement_horizontal_bgr;
+        break;
+
+    case WL_OUTPUT_SUBPIXEL_VERTICAL_RGB:
+        dcout.subpixel_arrangement = MirSubpixelArrangement::mir_subpixel_arrangement_vertical_rgb;
+        break;
+
+    case WL_OUTPUT_SUBPIXEL_VERTICAL_BGR:
+        dcout.subpixel_arrangement = MirSubpixelArrangement::mir_subpixel_arrangement_vertical_bgr;
+        break;
+
+    default:
+        dcout.subpixel_arrangement = MirSubpixelArrangement::mir_subpixel_arrangement_unknown;
+        break;
+    }
+}
+
+
+void mgw::DisplayClient::Output::mode(
+    void *data,
+    struct wl_output* /*wl_output*/,
+    uint32_t flags,
+    int32_t width,
+    int32_t height,
+    int32_t refresh)
+{
+    auto const output = static_cast<Output*>(data);
+    auto& dcout = output->dcout;
+    auto& modes = dcout.modes;
+
+    auto const mode = DisplayConfigurationMode{{width, height}, refresh/100000.0};
+    bool const is_current   = flags & WL_OUTPUT_MODE_CURRENT;
+    bool const is_preferred = flags & WL_OUTPUT_MODE_PREFERRED;
+
+    auto const found = std::find_if(begin(modes), end(modes), [&](auto const& m)
+        { return mode.size == m.size && mode.vrefresh_hz == m.vrefresh_hz; });
+
+    if (found != end(modes))
+    {
+        if (is_current) dcout.current_mode_index = distance(begin(modes), found);
+        if (is_preferred) dcout.preferred_mode_index = distance(begin(modes), found);
+    }
+    else
+    {
+        modes.push_back(mode);
+        if (is_current) dcout.current_mode_index = modes.size() - 1;
+        if (is_preferred) dcout.preferred_mode_index = modes.size() - 1;
+    }
+}
+
+void mgw::DisplayClient::Output::scale(void* /*data*/, wl_output* /*wl_output*/, int32_t /*factor*/)
+{
+}
+
+mgw::DisplayClient::Output::Output(
+    wl_output* output,
+    DisplayClient* owner,
+    std::function<void(Output const&)> on_constructed,
+    std::function<void(Output const&)> on_change) :
+    output{output},
+    owner{owner},
+    surface{wl_compositor_create_surface(owner->compositor)},
+    on_done{[this, on_constructed = std::move(on_constructed), on_change=std::move(on_change)]
+        (Output const& o) mutable { on_constructed(o), on_done = std::move(on_change); }}
```

Comment:
What happens here when `on_done()` is called the third time?

Reply:
> What happens here when `on_done()` is called the third time?

Unless I'm missing something, the supplied `on_change` handler is called for the second time.

Reply:
...oh, yes!

Is it safe to overwrite the `std::function` from its `operator()`?

Reply:
Same rules as `delete this`: Do not access its state afterwards.